### PR TITLE
properly test build chain call budget

### PR DESCRIPTION
**Pull request by cpu**
Previously (https://github.com/rustls/webpki/pull/168) we fixed the test case  that used the `build_degenerate_chain` helper to test the build chain call budget (introduced in https://github.com/rustls/webpki/pull/163) by artificially inflating the signature validation limit. This was crummy for a few reasons:

1. It required faking the budget, both to increase the signature validation limit (to avoid hitting that limit first) and to decrease the build chain call budget (to avoid the test taking ~6s to run).
2. It didn't demonstrate very well why two separate limits are required, as it gave the sense under normal circumstances the signature validation limit would reject the pathological chain.

This branch reworks `build_degenerate_chain` so that it can reproduce a pathological chain that only has quadratic runtime prevented by the build chain call budget. E.g. the signature validation limit is _not_ sufficient for this test case. The test demonstrates this with the default budget for both signature validations and build chain calls.

Along the way I did some small refactoring of the `test_utils` to make it easier to adjust certificate parameters for tests that require that without burdening tests that don't.

### CODE REVIEW

**Comment by cpu** [comment]
For fun (and a warm laptop) you can apply this commit to see the quadratic runtime when the build chain budget is neutered and the pathological chain size increased a little bit for maximum effect: https://github.com/cpu/webpki/commit/3584ed4f524553ded37cd671cb76036a245222dc

File: src/test_utils.rs [hunk 1]
```diff
@@ -22,8 +22,16 @@ pub(crate) fn make_issuer(
         rcgen::KeyUsagePurpose::CrlSign,
     ];
     ca_params.alg = RCGEN_SIGNATURE_ALG;
-    ca_params.name_constraints = name_constraints;
-    rcgen::Certificate::from_params(ca_params).unwrap()
+    ca_params
+}
+
+pub(crate) fn make_issuer(
```
Comment (djc) [inline src/test_utils.rs]: Nit: so change the order? :)

File: src/test_utils.rs [hunk 2]
```diff
@@ -29,12 +29,16 @@ pub(crate) fn make_issuer(org_name: impl Into<String>) -> rcgen::Certificate {
...
-pub(crate) fn make_end_entity(issuer: &rcgen::Certificate) -> CertificateDer<'static> {
-    let mut ee_params = rcgen::CertificateParams::new(vec!["example.com".to_string()]);
+pub(crate) fn end_entity_params(subject_alt_names: Vec<String>) -> rcgen::CertificateParams {
+    let mut ee_params = rcgen::CertificateParams::new(subject_alt_names);
     ee_params.is_ca = rcgen::IsCa::ExplicitNoCa;
     ee_params.alg = RCGEN_SIGNATURE_ALG;
+    ee_params
+}
+
+pub(crate) fn make_end_entity(issuer: &rcgen::Certificate) -> CertificateDer<'static> {
```
Comment (djc) [inline src/test_utils.rs:37]: Same nit: order?

File: src/verify_cert.rs [hunk 3]
```diff
@@ -508,21 +508,25 @@ mod tests {
     }
 
     #[cfg(feature = "alloc")]
-    enum TrustAnchorIsActualIssuer {
-        Yes,
-        No,
+    enum ChainTrustAnchor {
```
Comment (djc) [inline src/verify_cert.rs:511]: Maybe this should be a separate commit?

**Review by djc — APPROVED** [review]

File: src/test_utils.rs (hunk 1 above)
Comment (cpu) [reply src/test_utils.rs]: Done.

File: src/test_utils.rs (hunk 2 above)
Comment (cpu) [reply src/test_utils.rs:30]: Done.

File: src/verify_cert.rs (hunk 3 above)
Comment (cpu) [reply src/verify_cert.rs:511]: Sure, separated out :+1:

**Comment by cpu** [comment]
Going to administratively merge this while the nightly docs & coverage CI is borked pending upstream updates.
